Promises support

piranna opened the pull request:
Add support for `fs.promises`. It also add support for the newer functions not available before. The pull-request is not bacward compatible due to removal of `SyncWriteStream` function (seems a typo for me since I was not able to find it in the docs, please confirm and I'll re-add it if needed), and fix resolution of up dir to work the same way as `..` direntry in root dir, that in POSIX systems it points to itself, so `/../blah` is the same as `/blah`.

Please tell me any concerns do you have :-)

juliangruber commented on index.js:
```diff
@@ -1,26 +1,45 @@
 
+
+const ONE_ARGUMENT_PROMISES =
+[
```
could you please remove those newlines?

juliangruber replied:
let me add standard to this :)

juliangruber replied:
Could you please run `npx prettier-standard '**/*.js'` on your patch? That should make it mergeable with current master.

piranna replied:
Done :-)

piranna commented:
There was still some conflicts, now they are fixed.

juliangruber commented:
This looks great, thank you! Would you mind also adding some tests, so we can verify this doesn't break in the future? If not that is ok also and I can merge as is.

piranna commented:
> This looks great, thank you!

You are welcome :-)


> Would you mind also adding some tests, so we can verify this doesn't break in the future? If not that is ok also and I can merge as is.

No problem. Can you be able to configure Travis CI and Coveralls while I'm adding them? :-)

juliangruber commented:
Travis is set up already. Coveralls I've never done, if I add you as an owner on this repo, could you do it? And would you be interested in maintaining this in general?

piranna commented:
> Travis is set up already

But... it's not checking the pull-requests :-/ Maybe you need to enable that feature?


> Coveralls I've never done, if I add you as an owner on this repo, could you do it?

Yes, I think there would not be problems.


> And would you be interested in maintaining this in general?

Maybe :-) Main problem is publishing to npm, besides that I would not have so much trouble having another child to feed :-D

juliangruber commented:
> But... it's not checking the pull-requests :-/ Maybe you need to enable that feature?

It should run on the next commit, travis was disabled for some reason.

> Yes, I think there would not be problems.

Invited you :)

> Maybe :-) Main problem is publishing to npm, besides that I would not have so much trouble having another child to feed :-D

I can add you to both!

piranna commented:
> > But... it's not checking the pull-requests :-/ Maybe you need to enable that feature?
> 
> It should run on the next commit, travis was disabled for some reason.
> 
> >

Nice.


> > Yes, I think there would not be problems.
> 
> Invited you :)

Thank you! :-)


> > Maybe :-) Main problem is publishing to npm, besides that I would not have so much trouble having another child to feed :-D
> 
> I can add you to both!

That would be cool :-)

By the way, [I have just enabled Coveralls for the repo](https://coveralls.io/github/juliangruber/subfs), just only last to generate coverage info and send it.

piranna commented:
I have updated the code to use ES8 features and optimize the resolve algorythm. I've also changed the API to make the filesystem optional using the Node.js`fs` module as default value, and added a `_resolve()` function bound to the used base dir.

Note: tests are not passing because you've configured Travis to use Node.js 0.10, not Node.js 10 :-P

juliangruber commented:
> Note: tests are not passing because you've configured Travis to use Node.js 0.10, not Node.js 10 :-P

Just change the travis.yml, we'll release a major after this

piranna commented:
Done, tests are passing. Should we update the tests to use `jest` so we can have code coverage by free, or is it good enough just like how tests are at this moment?

juliangruber commented:
do as you wish :) either is ok for me

piranna commented:
Tests converted to Jest :-)

piranna commented:
> nice work!

Thank you! :-D
